<a href="https://colab.research.google.com/github/GUNAPILLCO/neural_profit/blob/main/notebooks/stage_03_target_definition/stage_03_target_definition.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **stage_03_target_definition**


## **Configuración del Entorno**


### 0.1. Acceso a Drive

In [29]:
from google.colab import drive
drive.mount('/content/drive')
drive_path = "/content/drive/MyDrive/neural_profit"

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### 0.2. Importación de librerías


In [30]:
import sys
import re
#Instalación de librería pandas_market_calendars
#!{sys.executable} -m pip install -q pandas_market_calendars
#print("Librería instalada: pandas_market_calendars")


from functools import reduce
# Utilidades generales
from datetime import datetime, timedelta
import os
import glob
import requests
import warnings
warnings.filterwarnings('ignore')

# Manejo y procesamiento de datos
import ta
import pandas as pd
import numpy as np
from tabulate import tabulate
import matplotlib.pyplot as plt
# Calendario de mercados
#import pandas_market_calendars as mcal

from ta.momentum import StochasticOscillator, ROCIndicator
from ta.volatility import BollingerBands, AverageTrueRange

from scipy.stats import spearmanr
from __future__ import annotations  # permite type hints modernos (Python < 3.11)
import json                         # para guardar el summary como JSON
import os                           # para leer variables de entorno
from pathlib import Path            # manejo robusto de rutas
import pandas as pd

### 0.3. Definición de rutas

In [31]:
DRIVE_DIR =   Path(os.environ.get("DRIVE_DIR", "/content/drive/MyDrive/neural_profit/"))

In [32]:
# ---------------------------------------------------------------------
# Configuración de rutas (DVC-friendly)
# ---------------------------------------------------------------------
# Estas rutas son RELATIVAS al repositorio.
# DVC necesita paths locales y determinísticos.
# Si mañana quiere apuntar a Drive, se hace vía DVC remote o symlink,
# NO cambiando la lógica del stage.

# PARA EL DVC:
IN_PARQUET = Path(os.environ.get("IN_PARQUET", "data/processed/mnq_intraday.parquet"))
OUT_PARQUET = Path(os.environ.get("OUT_PARQUET", "data/processed/mnq_intraday_labeled.parquet"))
OUT_SUMMARY = Path(os.environ.get("OUT_SUMMARY", "reports/target_definitio_summary.json"))

In [33]:
# PARA EL NOTEBOOK:
IN_PARQUET = DRIVE_DIR / IN_PARQUET
OUT_PARQUET = DRIVE_DIR / OUT_PARQUET
OUT_SUMMARY = DRIVE_DIR / OUT_SUMMARY

## **1. Definición de parámetros de operación**

El objetivo de este proyecto es la obtención de un rédito económico mediante la operatoria sobre el índice MNQ.

El plan de operación propuesto se define bajo los siguientes parámetros:

- Instrumento: MNQ
- Frecuencia operativa: hasta 4 operaciones diarias
- Meta diaria: mínimo USD 200, con un objetivo ideal de hasta USD 500
- Valor por punto del MNQ: USD 2

Esto se traduce en los siguientes objetivos diarios expresados en puntos:

- USD 200 ⇒ 100 puntos diarios
- USD 500 ⇒ 250 puntos diarios

Considerando un máximo de 4 operaciones por día, el objetivo promedio por operación resulta:

- Objetivo mínimo: 25 puntos por trade (100 / 4)
- Objetivo ideal: 62,5 puntos por trade (250 / 4)

Bajo esta premisa, es necesario evaluar la relación entre los retornos logarítmicos y la variación en puntos del índice, así como la frecuencia con la que se producen movimientos de esta magnitud.

## **2. Evaluación de feature de entrada `close`**




### 2.1. Justificación de análisis

A partir de la definición de los parámetros de operación establecidos en el punto anterior, resulta necesario analizar la variable `close`, la cual representa el nivel del índice MNQ en cada instante temporal del dataset intradía.

La evaluación de esta variable constituye un paso previo e indispensable al análisis estadístico de los retornos, por los siguientes motivos:

---

**1. Relación entre objetivos económicos y nivel del índice**

Los objetivos definidos en el Punto 1 se expresan en puntos del índice (25 a 62,5 puntos por operación). Sin embargo, los retornos utilizados en el modelado se calcularán en forma logarítmica, según la expresión:

$$
    r_t = \ln\left( \frac{close_{t+h}}{close_t} \right)
$$

Esto implica que un mismo desplazamiento en puntos absolutos del índice no se traduce en un retorno constante, sino que depende directamente del valor de `close_t`. Por lo tanto, para poder relacionar correctamente los objetivos económicos definidos en puntos con los retornos logarítmicos, es imprescindible conocer el orden de magnitud y la variabilidad del nivel del índice.

---

**2. Justificación del uso de close como referencia de escala**

La variable `close` actúa como factor de escala entre el retorno logarítmico y la variación absoluta en puntos del MNQ. En términos prácticos, la conversión puede aproximarse como:

$$
\Delta \text{puntos} \approx r_t \times close_t
$$

En consecuencia:

- definir umbrales de retorno sin considerar `close` conduce a criterios arbitrarios

- analizar `close` permite establecer umbrales de retorno dinámicos, coherentes con distintos niveles del índice.

---

**3. Necesidad de una referencia común antes del análisis estadístico**

Dado que el dataset abarca múltiples períodos y regímenes de mercado, el nivel del índice MNQ presenta variaciones significativas a lo largo del tiempo. Por este motivo, antes de evaluar la frecuencia y distribución de los retornos, es metodológicamente correcto:

  1. Comprender la escala real del índice representada por `close`.
  2. Validar que los objetivos definidos en puntos sean razonables en todo el período analizado.
  3. Evitar sesgos derivados de asumir un nivel de precios constante.

---

**4 Rol de este paso dentro del pipeline**

La evaluación de la variable `close` cumple, dentro del pipeline, la función de:

- Conectar los objetivos económicos con las variables financieras del dataset.
- Establecer una base sólida para la posterior evaluación estadística de los retornos.
- Garantizar coherencia entre la formulación del problema y la realidad operativa.

---

En síntesis, el análisis de la feature `close` no persigue inicialmente fines estadísticos, sino que constituye un paso conceptual y metodológico orientado a asegurar que los objetivos económicos definidos sean correctamente traducidos al lenguaje de los retornos financieros. Solo a partir de esta validación resulta pertinente avanzar al análisis estadístico de los retornos y a la definición final de los targets de predicción.

### 2.2. Aplicación de análisis

#### 2.2.1. Carga de dataset `intraday_mnq`


In [34]:
def load_mnq_parquet():
    os.path.exists(IN_PARQUET)
    print("Archivo encontrado en disco. Cargando dataset local...")
    mnq_parquet = pd.read_parquet(IN_PARQUET)
    return mnq_parquet

In [35]:
def add_column_date(df):
    # Asegurar que el índice esté en formato datetime
    df.index = pd.to_datetime(df.index)

    # Crear una nueva columna 'date' con la fecha extraída del índice
    df['date'] = df.index.date

    # Reordenar columnas: 'date', 'time_str', y luego el resto
    cols = ['date'] + [col for col in df.columns if col not in ['date']]

    df = df[cols]

    return df

In [36]:
def info_dataset(df):
  print("Información del dataset:\n")

  # Contar valores únicos en la columna 'date'
  num_dias = df['date'].nunique()
  print(f"\tCantidad de días: {num_dias}")

  # Filtrar valores válidos
  validos_por_dia = df.dropna(subset=['close']).groupby('date').size()

  # Calcular el promedio
  promedio_por_fecha = validos_por_dia.mean()
  print(f"\tRegistros por día: {int(promedio_por_fecha)}")

  primer_hora = df.index[0].strftime('%H:%M')
  ultima_hora = df.index[-1].strftime('%H:%M')
  zona_horaria = df.index[0].tzinfo


  print(f"\tHora diaria de inicio {primer_hora}")
  print(f"\tHora diaria de final {ultima_hora}")
  print(f"\tZona horaria: {zona_horaria}")

In [61]:
mnq_intraday = load_mnq_parquet()
mnq_intraday = add_column_date(mnq_intraday)
mnq_intraday.head()


Archivo encontrado en disco. Cargando dataset local...


,date,open,high,low,close,volume
datetime,,,,,,
2019-12-23 06:30:00-05:00,2019-12-23,8727.75,8728.00,8727.75,8727.75,7
2019-12-23 06:31:00-05:00,2019-12-23,8727.50,8727.75,8726.50,8726.50,89
2019-12-23 06:32:00-05:00,2019-12-23,8726.50,8726.50,8725.25,8725.25,34
2019-12-23 06:33:00-05:00,2019-12-23,8725.50,8726.25,8724.75,8726.00,53
2019-12-23 06:34:00-05:00,2019-12-23,8726.00,8726.00,8726.00,8726.00,3


In [62]:
info_dataset(mnq_intraday)

Información del dataset:

	Cantidad de días: 1303
	Registros por día: 571
	Hora diaria de inicio 06:30
	Hora diaria de final 16:00
	Zona horaria: America/New_York


#### 2.2.2. Análisis estadistico de `close`


In [38]:
import pandas as pd
from typing import Dict, Tuple

def analyze_close_statistics(
    df: pd.DataFrame,
    close_col: str = "close",
    percentiles=(0.01, 0.05, 0.25, 0.50, 0.75, 0.95, 0.99),
    by_year: bool = True
) -> Tuple[pd.Series, Dict[str, float], pd.DataFrame | None]:
    """
    Calcula estadísticas descriptivas de la variable close.

    Parámetros
    ----------
    df : pd.DataFrame
        DataFrame con índice datetime y columna close.
    close_col : str
        Nombre de la columna de precios (default: 'close').
    percentiles : tuple
        Percentiles a calcular.
    by_year : bool
        Si True, devuelve estadísticas agregadas por año.

    Retorna
    -------
    close_describe : pd.Series
        Resultado de df[close].describe(percentiles=...)
    summary : dict
        Resumen compacto con métricas clave.
    close_by_year : pd.DataFrame | None
        Estadísticas por año (o None si by_year=False).
    """

    close_series = df[close_col].dropna()

    # 1) Estadísticas globales
    close_describe = close_series.describe(percentiles=percentiles)

    # 2) Resumen compacto
    summary = {
        "count": int(close_series.count()),
        "mean": float(close_series.mean()),
        "std": float(close_series.std()),
        "min": float(close_series.min()),
        "p01": float(close_series.quantile(0.01)),
        "p05": float(close_series.quantile(0.05)),
        "p50": float(close_series.quantile(0.50)),
        "p95": float(close_series.quantile(0.95)),
        "p99": float(close_series.quantile(0.99)),
        "max": float(close_series.max()),
    }

    # 3) Estadísticas por año (opcional)
    close_by_year = None
    if by_year:
        tmp = df[[close_col]].copy()
        tmp["year"] = tmp.index.year

        close_by_year = tmp.groupby("year")[close_col].agg(
            count="count",
            mean="mean",
            median="median",
            p05=lambda s: s.quantile(0.05),
            p95=lambda s: s.quantile(0.95),
            min="min",
            max="max",
        )

    return close_describe, summary, close_by_year


### 2.2. Evaluación de resultados

In [66]:
close_desc, close_summary, close_yearly = analyze_close_statistics(mnq_intraday)

print(close_desc)
#print(close_summary)
#print(close_yearly)


count    744013.000000
mean      14762.034275
std        3566.973264
min        6765.750000
1%         8099.780000
5%         9110.500000
25%       12065.250000
50%       14430.750000
75%       17513.000000
95%       21284.000000
99%       21908.720000
max       22317.250000
Name: close, dtype: float64


Algunas conclusiones del análisis de la variable `close`:


1. Nivel típico del índice MNQ

    La media de la variable `close` se ubica en 14 762 puntos, con una mediana cercana (14 431 puntos).
    Esto indica que, a lo largo de todo el período analizado, el MNQ ha operado la mayor parte del tiempo en torno al nivel de 15 000 puntos, validando dicho valor como referencia central de escala.

2. Amplio rango de valores y múltiples regímenes

    El rango observado va desde 6 766 hasta 22 317 puntos, lo que evidencia:

    - La presencia de múltiples regímenes de mercado
    - Una evolución estructural del índice a lo largo del tiempo
    - La no estacionariedad del nivel de precios.

    Este comportamiento descarta el uso de un único valor fijo de referencia para todo el período.

3. Concentración de observaciones en un rango operativo claro

    El 50 % central de los datos (P25–P75) se encuentra entre 12 065 y 17 513 puntos, mientras que el 90 % de las observaciones (P5–P95) se concentra aproximadamente entre 9 110 y 21 284 puntos.

    Esto define un rango operativo predominante, dentro del cual se desarrollan la mayoría de las oportunidades intradía.

4. Implicancia directa sobre la conversión puntos ↔ retornos

    Dado que el mismo desplazamiento en puntos genera retornos distintos según el nivel de `close`, la variabilidad observada implica que:

    - Un objetivo fijo en puntos (ej. 25 o 60 puntos) no corresponde a un retorno fijo.
    - Los umbrales de retorno deben ser dinámicos y dependientes de close_t.

    Esta conclusión es central para evitar sesgos de escala en la definición posterior de targets.

5. Consistencia con el objetivo económico del proyecto

    El rango y la media observados son coherentes con los objetivos definidos en el Punto 1, ya que:

    - Los niveles típicos del índice permiten que movimientos de 25 a 60 puntos representen retornos intradía realistas.

    -Dichos movimientos no corresponden a eventos extremos en la mayor parte del período analizado.

**Conclusión general**

El análisis de la variable close confirma que:

- El nivel del índice MNQ presenta una variabilidad significativa pero acotada,
- La media cercana a 15 000 puntos es representativa a nivel global,
- Cualquier definición de retornos objetivo debe considerar close como factor de escala dinámico.

Con estas conclusiones establecidas, el siguiente paso lógico es evaluar cómo se comportan los retornos logarítmicos en relación con estos niveles de precio, y con qué frecuencia permiten alcanzar los objetivos económicos definidos.

## **3. Evaluación de los retornos logarítmicos**


Una vez establecidos los parámetros económicos de la operatoria (Punto 1) y analizado el nivel y la variabilidad del índice MNQ a través de la variable `close` (Punto 2), corresponde evaluar el comportamiento estadístico de los retornos logarítmicos, con el objetivo de determinar si los movimientos necesarios para alcanzar los objetivos económicos definidos ocurren con una frecuencia razonable.

Para este análisis se consideran los retornos acumulados a distintos horizontes temporales (`ret_30`, `ret_60`, `ret_90` y `ret_120`), los cuales representan la variación relativa del precio del índice en ventanas de 30, 60, 90 y 120 minutos, respectivamente.

### **3.1. Justificación del uso de retornos**

El análisis del movimiento del índice MNQ no se realiza directamente sobre el precio (`close`), sino sobre sus retornos logarítmicos, debido a razones metodológicas y financieras bien establecidas.

En particular, los retornos logarítmicos:

- permiten comparar movimientos de precios en distintos niveles del índice,
- son aditivos en el tiempo, lo que facilita el análisis en ventanas temporales,
- presentan propiedades estadísticas más estables que los precios absolutos,
- constituyen la forma estándar de modelar variaciones relativas en finanzas cuantitativas.

Dado que el objetivo del proyecto es evaluar movimientos relativos del mercado en horizontes intradía, el retorno logarítmico resulta una representación más adecuada que la variación absoluta del precio.

### **3.2 Definición y cálculo de los retornos**

El retorno logarítmico se define como:

$$
r_t = \ln\left( \frac{close_{t+h}}{close_t} \right)
$$

donde:

- `close_t` es el valor del índice en el instante actual,
- `close_{t+h}` es el valor del índice luego de un horizonte temporal
`h`.


En este proyecto se consideran retornos acumulados en cuatro horizontes:

- `ret_30`: retorno a 30 minutos
- `ret_60`: retorno a 60 minutos
- `ret_90`: retorno a 90 minutos
- `ret_120`: retorno a 120 minutos

Estos retornos representan la variación relativa del índice en ventanas temporales alineadas con la operatoria intradía planteada en el Punto 1.

#### **3.2.1. Cálculo de retornos por horizonte temporal**

In [72]:
def add_log_return(df):
    df['ret_30'] = df.groupby('date')['close'].transform(
        lambda x: np.log(x.shift(-30)) - np.log(x)
    )

    df['ret_60'] = df.groupby('date')['close'].transform(
        lambda x: np.log(x.shift(-60)) - np.log(x)
    )

    df['ret_90'] = df.groupby('date')['close'].transform(
        lambda x: np.log(x.shift(-90)) - np.log(x)
    )

    df['ret_120'] = df.groupby('date')['close'].transform(
        lambda x: np.log(x.shift(-120)) - np.log(x)
    )

    return df

In [73]:
mnq_intraday_with_returns = add_log_return(mnq_intraday)

In [74]:
mnq_intraday_with_returns.head()

,date,open,high,low,close,volume,ret_30,ret_60,ret_90,ret_120
datetime,,,,,,,,,,
2019-12-23 06:30:00-05:00,2019-12-23,8727.75,8728.00,8727.75,8727.75,7,0.000029,0.001031,0.000687,0.001059
2019-12-23 06:31:00-05:00,2019-12-23,8727.50,8727.75,8726.50,8726.50,89,0.000057,0.001059,0.000859,0.001145
2019-12-23 06:32:00-05:00,2019-12-23,8726.50,8726.50,8725.25,8725.25,34,0.000315,0.001117,0.000916,0.001231
2019-12-23 06:33:00-05:00,2019-12-23,8725.50,8726.25,8724.75,8726.00,53,0.000258,0.000974,0.000916,0.001231
2019-12-23 06:34:00-05:00,2019-12-23,8726.00,8726.00,8726.00,8726.00,3,0.000315,0.000916,0.000888,0.001346


### **3.3. Análisis descriptivo de retornos logarítmicos por horizonte temporal**

#### **3.3.1. Código de aplicación**

In [75]:
returns = ['ret_30', 'ret_60', 'ret_90', 'ret_120']

In [77]:
returns_stats = (
    mnq_intraday_with_returns[returns]
    .dropna()
    .describe(percentiles=[0.01, 0.05, 0.5, 0.95, 0.99])
)

#### **3.3.2. Análisis y Conclusiones**

In [78]:
returns_stats

,ret_30,ret_60,ret_90,ret_120
count,587653.000000,587653.000000,587653.000000,587653.000000
mean,0.000016,0.000039,0.000071,0.000100
std,0.003010,0.004274,0.005276,0.006141
min,-0.053377,-0.056549,-0.049889,-0.057965
1%,-0.008783,-0.012448,-0.015042,-0.017120
5%,-0.004639,-0.006776,-0.008475,-0.009999
50%,0.000098,0.000185,0.000243,0.000330
95%,0.004292,0.006151,0.007751,0.009118
99%,0.008058,0.011347,0.013780,0.015915
max,0.082122,0.079896,0.083184,0.090994


1. Coherencia estadística entre horizontes

    Se observa un comportamiento consistente y esperable:
    - La media del retorno aumenta al ampliar el horizonte temporal.
    - El desvío estándar crece de forma monótona (30 → 120 min).

    Esto confirma que:
    - los retornos están correctamente calculados,
    - ventanas más largas capturan mayor acumulación de movimiento.

2. Media cercana a cero (propiedad intradía)

    En todos los horizontes, la media del retorno es muy próxima a cero:
    - no existe sesgo direccional sistemático,
    - el mercado intradía es esencialmente balanceado.

    Conclusión metodológica:
    - no es apropiado modelar retornos esperando una tendencia promedio,
    - el valor está en eventos específicos, no en el promedio.

3. Distribuciones aproximadamente simétricas

    Los percentiles negativos y positivos son de magnitud comparable:
    - colas negativas y positivas similares,
    - posibilidad real de movimientos alcistas y bajistas.

    Esto habilita:
    - modelos simétricos long/short,
    - formulaciones de clasificación direccional más adelante.

4) Incremento de magnitud con el horizonte

    Los percentiles altos crecen de forma clara:

    P95:
    - ret_30 ≈ 0.43 %
    - ret_60 ≈ 0.62 %
    - ret_90 ≈ 0.78 %
    - ret_120 ≈ 0.91 %

    P99:
    - ret_30 ≈ 0.81 %
    - ret_120 ≈ 1.59 %

    Conclusión:
    - los movimientos económicamente relevantes existen,
    - pero su frecuencia depende fuertemente del horizonte temporal.

5. Presencia de colas extremas

    Los valores mínimos y máximos muestran:
    - retornos extremos (±5 % a ±9 %),
    - asociados a eventos excepcionales (aperturas, noticias, shocks).

    Implicación:
    - estos eventos no deben usarse como referencia operativa base,
    - pero confirman que el dataset captura escenarios de estrés reales.

6. Rol de cada horizonte en el pipeline

    A partir del análisis (sin decidir targets aún):
    - 30 min: muy frecuente, pero movimientos pequeños.
    - 60 min: buen equilibrio entre frecuencia y amplitud.
    - 90–120 min: movimientos grandes, menor frecuencia, escenarios de continuidad.

    Esto permite, en el siguiente paso, alinear nuestros objetivos económicos con horizontes, sin forzar supuestos.



**Conclusión general**

El análisis descriptivo de los retornos logarítmicos muestra que:

- el mercado del MNQ presenta movimientos intradía de magnitud creciente con el horizonte,
- los retornos tienen propiedades estadísticas sanas y coherentes,
- existen movimientos suficientes para sustentar objetivos económicos razonables,
- pero no de forma uniforme ni constante.

Con estas conclusiones, el siguiente paso lógico es evaluar qué magnitudes de retorno son compatibles con los objetivos económicos definidos, y recién entonces definir los targets.

## **4. Vinculación entre nivel de precio (`close`), retornos y objetivos económicos**

### **4.1. Explicación metodológica**

El objetivo de este punto es integrar los resultados obtenidos en los Puntos 2 y 3 con los parámetros económicos definidos en el Punto 1, de modo de establecer una relación cuantitativa coherente entre:

- el nivel del índice MNQ (`close`),
- los retornos logarítmicos (`ret_**`) intradía,
- y los objetivos económicos expresados en puntos y dólares.

Esta vinculación es un paso intermedio indispensable antes de definir formalmente los targets de predicción.

**1. Principio de correspondencia entre puntos y retornos**

  Los objetivos económicos del proyecto se formulan en términos de variación absoluta del índice (puntos), mientras que el comportamiento estadístico del mercado se analiza mediante retornos logarítmicos. Para conectar ambos dominios se utiliza la relación:

  $$
    \Delta \text{puntos} \approx r_{t,h} \times close_t
  $$
      
  donde:

  - `𝑟_{𝑡,ℎ}` es el retorno logarítmico acumulado en un horizonte `ℎ`
  - `𝑐𝑙𝑜𝑠𝑒_𝑡` es el nivel del índice al inicio de la ventana.

  Esta expresión permite traducir cualquier retorno observado a una variación en puntos directamente interpretable desde el punto de vista operativo.
<br><br>
**2. Uso del nivel de precio como factor de escala dinámico**

Dado que el análisis de la variable `close` evidenció una variabilidad significativa del nivel del índice a lo largo del tiempo, la conversión entre retornos y puntos no puede basarse en un valor fijo de referencia.

Por el contrario:
- cada observación debe evaluarse en función de su propio `close_t`,
- los umbrales de retorno asociados a un objetivo en puntos se definen de forma dinámica y dependiente del nivel del mercado.

Este enfoque garantiza que:
- los criterios operativos sean coherentes en distintos regímenes de precio,
- los resultados no estén sesgados hacia períodos específicos del dataset.
<br><br>
**3. Integración de horizontes temporales y objetivos operativos**

La vinculación entre precio y retorno se realiza de manera conjunta con el horizonte temporal del movimiento, dado que:

- distintos horizontes presentan distintas combinaciones de frecuencia y magnitud,
- los objetivos económicos definidos requieren movimientos de cierta amplitud en ventanas temporales compatibles con la operatoria intradía.

En consecuencia, esta etapa no busca identificar un único horizonte óptimo, sino evaluar cómo cada horizonte contribuye al cumplimiento de los objetivos económicos, considerando tanto la magnitud del movimiento como su frecuencia histórica.
<br><br>
**4. Rol de esta vinculación en la definición de targets**

La finalidad de este punto no es aún fijar los targets, sino:

- determinar qué rangos de retorno son compatibles con los objetivos económicos definidos,
- identificar qué horizontes temporales concentran dichos movimientos,
- establecer una base cuantitativa objetiva para la definición posterior de los targets de predicción.
<br><br>
**Cierre del punto (explicativo)**

En síntesis, la vinculación entre el nivel de precio, los retornos logarítmicos y los objetivos económicos permite traducir los requerimientos operativos del proyecto al lenguaje estadístico del dataset, asegurando coherencia entre la formulación del problema, el comportamiento histórico del mercado y la realidad de la operatoria intradía.

### **4.2. Vinculación cuantitativa**

La vinculación cuantitativa se hace convirtiendo nuestros objetivos en puntos (definidos en el Punto 1) a umbrales de retorno logarítmico usando el nivel del índice (`close`) (Punto 2), y luego contrastándolos con la distribución de retornos (Punto 3).
<br><br>
**1. Conversión exacta (umbral dinámico)**

Para un objetivo de `Δ` puntos en un horizonte `ℎ`:
$$
r_{\text{umbral},t} \approx \ln\left(1 + \frac{\Delta}{close_t}\right) \approx \frac{\Delta}{close_t}
$$

En práctica intradía (movimientos pequeños), la aproximación `Δ/𝑐𝑙𝑜𝑠𝑒_𝑡` es suficiente y consistente con tus `ret_h`.
<br><br>
**2. Umbrales de retorno equivalentes usando percentiles de `close`**

Como close varía mucho (2019–2025), el retorno necesario para lograr los mismos puntos cambia. Tomamos 3 niveles representativos:

  - P05 close = 9 110.5
  - P50 close = 14 430.75
  - P95 close = 21 284.0

Objetivo mínimo: 25 puntos por trade

$$r_{25} \approx \frac{25}{close_t}$$

En P05: $$\frac{25}{9110.5} = 0.002744 \;\Rightarrow\; 0.274\%$$
En P50: $$\frac{25}{14430.75} = 0.001732 \;\Rightarrow\; 0.173\%$$
En P95: $$\frac{25}{21284} = 0.001174 \;\Rightarrow\; 0.117\%$$


Objetivo ideal: 62.5 puntos por trade

$$r_{62.5} \approx \frac{62.5}{close_t}$$
	​
En P05: $$\frac{62.5}{9110.5} = 0.006860 \;\Rightarrow\; 0.686\%$$
En P50: $$\frac{62.5}{14430.75} = 0.004331 \;\Rightarrow\; 0.433\%$$
En P95: $$ \frac{62.5}{21284} = 0.002936\;\Rightarrow\; 0.294\%$$

<br><br>

**Conclusión cuantitativa inmediata:**

- Lograr 25 pts requiere típicamente retornos del orden 0.12%–0.27% (según régimen de close).
- Lograr 62.5 pts requiere típicamente 0.29%–0.69%.

### 2.1. Justificación

### 2.2. Aplicación

In [56]:
mnq_intraday_labeled

,date,open,high,low,close,volume,ret_30,ret_60,ret_90
datetime,,,,,,,,,
2019-12-23 06:30:00-05:00,2019-12-23,8727.75,8728.00,8727.75,8727.75,7,0.000029,0.001031,0.000687
2019-12-23 06:31:00-05:00,2019-12-23,8727.50,8727.75,8726.50,8726.50,89,0.000057,0.001059,0.000859
2019-12-23 06:32:00-05:00,2019-12-23,8726.50,8726.50,8725.25,8725.25,34,0.000315,0.001117,0.000916
2019-12-23 06:33:00-05:00,2019-12-23,8725.50,8726.25,8724.75,8726.00,53,0.000258,0.000974,0.000916
2019-12-23 06:34:00-05:00,2019-12-23,8726.00,8726.00,8726.00,8726.00,3,0.000315,0.000916,0.000888
...,...,...,...,...,...,...,...,...,...
2025-06-13 15:56:00-04:00,2025-06-13,21624.50,21635.00,21613.50,21617.50,3251,NaN,NaN,NaN
2025-06-13 15:57:00-04:00,2025-06-13,21616.50,21635.25,21615.75,21623.75,2201,NaN,NaN,NaN
2025-06-13 15:58:00-04:00,2025-06-13,21623.25,21632.75,21616.50,21621.75,1859,NaN,NaN,NaN


## **3. Summary**

En esta etapa, el objetivo del summary es verificar y documentar que los targets generados:

- Existen y tienen la forma esperada.
- Presentan un porcentaje de NaN coherente con el horizonte.
- Tienen estadísticas básicas razonables (orden de magnitud, dispersión, outliers).

En la notebook:

- Se calculan y muestran métricas.
- No se guardan archivos (.json, .parquet).
- No se loguea en MLflow.

Eso se hará recién en el .py.

In [57]:
#Targets a analizar:


### 3.1. Porcentaje de NaNs por target

In [58]:
import pandas as pd

nan_summary = (
    mnq_intraday_labeled[targets]
    .isna()
    .mean()
    .mul(100)
    .to_frame(name='nan_pct')
)

nan_summary

,nan_pct
ret_30,5.253940
ret_60,10.507881
ret_90,15.761821


Qué se espera observar:

- ret_30 ≈ NaNs en los últimos 30 minutos de cada día.
- ret_60 ≈ NaNs en los últimos 60 minutos.
- ret_90 ≈ NaNs en los últimos 90 minutos.

Esto valida que el groupby(date) + shift(-h) funciona correctamente.

### 3.2. Estadísticas básicas de los targets (excluyendo NaNs)

,ret_30,ret_60,ret_90
count,626743.000000,626743.000000,626743.000000
mean,0.000027,0.000051,0.000076
std,0.002992,0.004264,0.005267
min,-0.053377,-0.056549,-0.049889
1%,-0.008700,-0.012411,-0.015035
5%,-0.004601,-0.006740,-0.008441
50%,0.000107,0.000189,0.000244
95%,0.004281,0.006147,0.007738
99%,0.008077,0.011407,0.013842
max,0.082122,0.079896,0.083184


Incluye:

- media
- desviación estándar
- percentiles (1%, 5%, mediana, 95%, 99%)
- mínimos y máximos

### 3.3. Estadísticas básicas de los targets (excluyendo NaNs)

## 4. Guardado de dataset de indicadores técnicos

Antes de comenzar, tengamos en cuenta que a partir de los resultados obtenidos, se observa que los indicadores técnicos no muestran capacidad predictiva consistente en horizontes cortos de 5 y 15 minutos.  
En estas fronteras, los valores de IC_mean son muy bajos y cercanos a cero, lo que indica ausencia de señal útil.  
Además, el aumento del IC_std confirma que cualquier correlación detectada en esos horizontes es poco estable y probablemente producto del ruido.  

Por este motivo, se eliminan las columnas correspondientes a target_return_5 y target_return_15, concentrando el análisis en horizontes de 30, 60 y 90 minutos, donde la mayoría de los indicadores comienzan a mostrar señales más claras y sostenidas.  

Además, en las conclusiones de cada indicador técnico ya habíamos señalado que indicadores técnicos conservaremos.
